File: mailblastr-cli/src/commands/contacts.js
'use strict';

const fs = require('node:fs');
const path = require('node:path');
const { CliError, parseJson, clean, withPagination, pagination } = require('../helpers');

/**
 * A contact lives either in a sending domain's pool (`--domain`, the
 * domain-first default) or in a plain audience (`--audience-id`, the nested
 * `/audiences/:id/contacts` API). Exactly one of the two names the container.
 */
function requireContainer(opts) {
  if (opts.domain && opts.audienceId) {
    throw new CliError('Provide only one of --domain or --audience-id.');
  }
  if (!opts.domain && !opts.audienceId) {
    throw new CliError('Provide --domain <domain> (or --audience-id <id> to target an audience).');
  }
}

function register({ group, leaf, act }) {
  const contacts = group('contacts', 'Manage contacts (domain-first: each sending domain has its own pool)');

  act(
    leaf(contacts, 'create', "Create a contact in a domain's pool")
      .option('--domain <domain>', 'the sending domain whose contact pool to use')
      .option('--audience-id <id>', 'create in this audience instead of a domain pool')
      .requiredOption('--email <email>', "the contact's email address")
      .option('--first-name <name>', 'first name')
      .option('--last-name <name>', 'last name')
      .option('--unsubscribed', 'create as unsubscribed')
      .option('--properties <json>', 'custom property values as a JSON object'),
    ({ client, opts }) => {
      requireContainer(opts);
      return client.contacts.create(
        clean({
          domain: opts.domain,
          audienceId: opts.audienceId,
          email: opts.email,
          first_name: opts.firstName,
          last_name: opts.lastName,
          unsubscribed: opts.unsubscribed,
          properties: parseJson(opts.properties, '--properties'),
        }),
      );
    },
  );

  act(
    withPagination(
      leaf(contacts, 'list', "List a domain's contacts")
        .option('--domain <domain>', 'the sending domain whose contact pool to list')
        .option('--audience-id <id>', 'list this audience instead of a domain pool')
        .option('--segment-id <id>', 'only contacts in this segment'),
    ),
    ({ client, opts }) => {
      requireContainer(opts);
      return client.contacts.list(
        clean({
          domain: opts.domain,
          audienceId: opts.audienceId,
          segment_id: opts.segmentId,
          ...pagination(opts),
        }),
      );
    },
  );

  act(
    leaf(contacts, 'get <id>', 'Retrieve a contact by id or email').option(
      '--domain <domain>',
      'disambiguates an email id across domain pools',
    ),
    ({ client, opts, args: [id] }) => client.contacts.get(clean({ id, domain: opts.domain })),
  );

  act(
    leaf(contacts, 'update <id>', 'Update a contact (id or email)')
      .option('--domain <domain>', 'disambiguates an email id across domain pools')
      .option('--first-name <name>', 'first name')
      .option('--last-name <name>', 'last name')
      .option('--unsubscribed', 'mark unsubscribed')
      .option('--subscribed', 'mark subscribed again')
      .option('--properties <json>', 'custom property values as a JSON object'),
    ({ client, opts, args: [id] }) =>
      client.contacts.update(
        clean({
          id,
          domain: opts.domain,
          first_name: opts.firstName,
          last_name: opts.lastName,
          unsubscribed: opts.unsubscribed ? true : opts.subscribed ? false : undefined,
          properties: parseJson(opts.properties, '--properties'),
        }),
      ),
  );

  act(
    leaf(contacts, 'delete <id>', 'Delete a contact (id or email)').option(
      '--domain <domain>',
      'disambiguates an email id across domain pools',
    ),
    ({ client, opts, args: [id] }) => client.contacts.remove(clean({ id, domain: opts.domain })),
  );

  act(
    leaf(contacts, 'add-to-segment <id> <segmentId>', 'Add a contact to a segment'),
    ({ client, args: [id, segmentId] }) => client.contacts.addToSegment(id, segmentId),
  );

  act(
    withPagination(leaf(contacts, 'segments <id>', 'List the segments a contact belongs to')),
    ({ client, opts, args: [id] }) => client.contacts.listSegments(id, pagination(opts)),
  );

  act(
    leaf(contacts, 'remove-from-segment <id> <segmentId>', 'Remove a contact from a segment'),
    ({ client, args: [id, segmentId] }) => client.contacts.removeFromSegment(id, segmentId),
  );

  act(
    withPagination(leaf(contacts, 'topics <id>', "Get a contact's topic subscriptions")),
    ({ client, opts, args: [id] }) => client.contacts.getTopics(id, pagination(opts)),
  );

  act(
    leaf(contacts, 'set-topics <id>', "Update a contact's topic subscriptions").requiredOption(
      '--topics <json>',
      'JSON array of {"id","subscription"} objects, e.g. \'[{"id":"top_1","subscription":"opt_in"}]\'',
    ),
    ({ client, opts, args: [id] }) =>
      client.contacts.updateTopics(id, { topics: parseJson(opts.topics, '--topics') }),
  );

  // Bulk import — from a JSON array of contacts, or a CSV file (SDK-3).
  act(
    leaf(contacts, 'batch <audienceId>', 'Bulk-import contacts from a JSON array (upsert by email; max 10,000)')
      .option('--file <path>', 'path to a JSON file containing an array of contact objects')
      .option('--data <json>', 'inline JSON array of contact objects')
      .option('--on-conflict <mode>', "how to resolve an existing email: 'upsert' (default) or 'skip'"),
    ({ client, opts, args: [audienceId] }) => {
      if (opts.file && opts.data) throw new CliError('Provide only one of --file or --data.');
      if (!opts.file && !opts.data) throw new CliError('Provide --file <path> or --data <json array>.');
      let contactsArr;
      if (opts.file) {
        let raw;
        try { raw = fs.readFileSync(opts.file, 'utf8'); }
        catch (err) { throw new CliError(`Cannot read --file ${opts.file}: ${err.message}`); }
        contactsArr = parseJson(raw, `--file ${opts.file}`);
      } else {
        contactsArr = parseJson(opts.data, '--data');
      }
      if (!Array.isArray(contactsArr)) throw new CliError('The contacts payload must be a JSON array.');
      return client.contacts.batch(clean({ audienceId, contacts: contactsArr, on_conflict: opts.onConflict }));
    },
  );

  act(
    leaf(contacts, 'import <audienceId>', 'Bulk-import contacts from a CSV file (email column required; header row optional)')
      .option('--csv <path>', 'path to a CSV file to send inline (max 5 MB / 10,000 rows)')
      .option('--storage-key <key>', 'key from `contacts import-upload`, once the file has been PUT to its upload URL')
      .option('--file-name <name>', 'name recorded for the archived source file (inline --csv mode)')
      .option('--segment-id <id>', 'also add every imported email to this segment (must belong to the audience)')
      .option('--on-conflict <mode>', "how to resolve an existing email: 'upsert' (default) or 'skip'")
      .option('--no-create-properties', 'strict mode: keep only columns matching a registered property'),
    ({ client, opts, args: [audienceId] }) => {
      if (opts.csv && opts.storageKey) {
        throw new CliError('Provide only one of --csv or --storage-key.');
      }
      if (!opts.csv && !opts.storageKey) {
        throw new CliError('Provide --csv <path> (or --storage-key <key> for an already-uploaded file).');
      }
      let csv;
      if (opts.csv) {
        try { csv = fs.readFileSync(opts.csv, 'utf8'); }
        catch (err) { throw new CliError(`Cannot read --csv ${opts.csv}: ${err.message}`); }
      }
      // commander sets opts.createProperties=false for --no-create-properties (default true).
      return client.contacts.import(clean({
        audienceId, csv, storage_key: opts.storageKey, file_name: opts.fileName,
        segment_id: opts.segmentId, on_conflict: opts.onConflict,
        create_properties: opts.createProperties === false ? false : undefined,
      }));
    },
  );

  act(
    leaf(
      contacts,
      'import-upload <audienceId>',
      'Mint a presigned upload URL for a CSV too large to send inline (up to 256 MB)',
    ).requiredOption('--csv <path>', 'path to the CSV file to upload (its name and size are sent, not its contents)'),
    ({ client, opts, args: [audienceId] }) => {
      let stat;
      try { stat = fs.statSync(opts.csv); }
      catch (err) { throw new CliError(`Cannot read --csv ${opts.csv}: ${err.message}`); }
      // PUT the file to the returned `upload_url`, then finish the import with
      // `contacts import <audienceId> --storage-key <storage_key>`.
      return client.contacts.createImportUpload({
        audienceId,
        filename: path.basename(opts.csv),
        size: stat.size,
      });
    },
  );
}

module.exports = { register };

File: mailblastr-dotnet/src/Mailblastr/Models/Contacts.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Mailblastr;

/// <summary>An audience. Domain-first model: each sending domain has one contact POOL audience.</summary>
public class Audience
{
    [JsonPropertyName("object")]
    public string Object { get; set; } = "audience";

    [JsonPropertyName("id")]
    public string Id { get; set; } = null!;

    [JsonPropertyName("name")]
    public string Name { get; set; } = null!;

    /// <summary>
    /// Set when this audience is a sending domain's contact POOL (one pool per
    /// domain, lazily created). Null on plain user-created audiences.
    /// </summary>
    [JsonPropertyName("domain")]
    public string? Domain { get; set; }

    [JsonPropertyName("created_at")]
    public string CreatedAt { get; set; } = null!;
}

/// <summary>Result of a Google-Sheet import (POST /audiences/:id/contacts/import-sheet).</summary>
public class SheetImportResult
{
    [JsonPropertyName("object")]
    public string Object { get; set; } = "sheet_import";

    [JsonPropertyName("imported")]
    public int Imported { get; set; }

    [JsonPropertyName("updated")]
    public int Updated { get; set; }

    [JsonPropertyName("skipped")]
    public int Skipped { get; set; }

    [JsonPropertyName("total")]
    public int Total { get; set; }

    [JsonPropertyName("segment_id")]
    public string SegmentId { get; set; } = null!;

    [JsonPropertyName("segment_name")]
    public string SegmentName { get; set; } = null!;

    [JsonPropertyName("segment_added")]
    public int SegmentAdded { get; set; }

    [JsonPropertyName("variables")]
    public List<string> Variables { get; set; } = new();
}

/// <summary>A contact.</summary>
public class Contact
{
    [JsonPropertyName("object")]
    public string Object { get; set; } = "contact";

    [JsonPropertyName("id")]
    public string Id { get; set; } = null!;

    [JsonPropertyName("email")]
    public string Email { get; set; } = null!;

    [JsonPropertyName("first_name")]
    public string? FirstName { get; set; }

    [JsonPropertyName("last_name")]
    public string? LastName { get; set; }

    [JsonPropertyName("unsubscribed")]
    public bool Unsubscribed { get; set; }

    /// <summary>Custom contact properties (own values merged over registered fallbacks).</summary>
    [JsonPropertyName("properties")]
    public Dictionary<string, JsonElement>? Properties { get; set; }

    [JsonPropertyName("created_at")]
    public string CreatedAt { get; set; } = null!;
}

/// <summary>
/// Payload for creating a contact. DOMAIN-FIRST: set <see cref="Domain"/> to
/// create in that sending domain's contact pool via the flat /contacts API
/// (REQUIRED there), or <see cref="AudienceId"/> to target a specific audience
/// via the nested API. The same address on two domains is two records with
/// separate consent.
/// </summary>
public class ContactCreateOptions
{
    /// <summary>
    /// The sending domain whose contact pool the contact lands in (e.g.
    /// <c>yourdomain.com</c>). REQUIRED whenever <see cref="AudienceId"/> is omitted.
    /// </summary>
    [JsonIgnore]
    public string? Domain { get; set; }

    /// <summary>Target a specific audience via the nested /audiences/:id/contacts API instead of <see cref="Domain"/>.</summary>
    [JsonIgnore]
    public string? AudienceId { get; set; }

    [JsonPropertyName("email")]
    public string Email { get; set; } = null!;

    [JsonPropertyName("first_name")]
    public string? FirstName { get; set; }

    [JsonPropertyName("last_name")]
    public string? LastName { get; set; }

    [JsonPropertyName("unsubscribed")]
    public bool? Unsubscribed { get; set; }

    /// <summary>Custom contact property values (string or number) keyed by the property key.</summary>
    [JsonPropertyName("properties")]
    public Dictionary<string, object?>? Properties { get; set; }
}

/// <summary>
/// Payload for updating a contact. <see cref="Id"/> is a contact id OR email.
/// On the flat API, set <see cref="Domain"/> when Id is an EMAIL
/// (disambiguates across pools). Set <see cref="AudienceId"/> to use the
/// nested audience API instead.
/// </summary>
public class ContactUpdateOptions
{
    /// <summary>Contact id OR email.</summary>
    [JsonIgnore]
    public string Id { get; set; } = null!;

    /// <summary>Disambiguates an EMAIL Id across domains (a contact id is exact and needs no domain).</summary>
    [JsonIgnore]
    public string? Domain { get; set; }

    /// <summary>Audience the contact belongs to. Omit for the flat /contacts/:id API.</summary>
    [JsonIgnore]
    public string? AudienceId { get; set; }

    [JsonPropertyName("first_name")]
    public string? FirstName { get; set; }

    [JsonPropertyName("last_name")]
    public string? LastName { get; set; }

    [JsonPropertyName("unsubscribed")]
    public bool? Unsubscribed { get; set; }

    [JsonPropertyName("properties")]
    public Dictionary<string, object?>? Properties { get; set; }
}

/// <summary>A single contact in a bulk import (the audience is on the call).</summary>
public class ContactInput
{
    [JsonPropertyName("email")]
    public string Email { get; set; } = null!;

    [JsonPropertyName("first_name")]
    public string? FirstName { get; set; }

    [JsonPropertyName("last_name")]
    public string? LastName { get; set; }

    [JsonPropertyName("unsubscribed")]
    public bool? Unsubscribed { get; set; }

    [JsonPropertyName("properties")]
    public Dictionary<string, object?>? Properties { get; set; }
}

/// <summary>Result of a batch / CSV contact import.</summary>
public class ImportContactsResponse
{
    [JsonPropertyName("object")]
    public string Object { get; set; } = "contact_import";

    /// <summary>Newly inserted.</summary>
    [JsonPropertyName("imported")]
    public int Imported { get; set; }

    /// <summary>Existing contacts updated.</summary>
    [JsonPropertyName("updated")]
    public int Updated { get; set; }

    /// <summary>Rows dropped: missing/invalid email OR left untouched under on_conflict 'skip'.</summary>
    [JsonPropertyName("skipped")]
    public int Skipped { get; set; }

    /// <summary>Distinct contacts processed.</summary>
    [JsonPropertyName("total")]
    public int Total { get; set; }

    /// <summary>CSV import only: rows without a usable email.</summary>
    [JsonPropertyName("invalid_rows")]
    public int? InvalidRows { get; set; }

    /// <summary>CSV import only: rows dropped because the plan's contact cap was reached.</summary>
    [JsonPropertyName("limit_skipped")]
    public int? LimitSkipped { get; set; }

    /// <summary>CSV import only: rows the importer itself rejected (suppressed/system addresses).</summary>
    [JsonPropertyName("system_skipped")]
    public int? SystemSkipped { get; set; }

    /// <summary>CSV import only: header columns that were not imported.</summary>
    [JsonPropertyName("ignored_columns")]
    public List<string>? IgnoredColumns { get; set; }

    /// <summary>CSV import only: emails added to <c>segment_id</c> when one was supplied.</summary>
    [JsonPropertyName("segment_added")]
    public int? SegmentAdded { get; set; }

    /// <summary>CSV import only: the archived source file (<c>file_name</c>, <c>storage_key</c>, <c>archived</c>).</summary>
    [JsonPropertyName("source_file")]
    public JsonElement? SourceFile { get; set; }

    /// <summary>CSV import only: plan/contact-cap accounting for the import.</summary>
    [JsonPropertyName("contact_limit")]
    public JsonElement? ContactLimit { get; set; }
}

/// <summary>
/// A presigned direct-to-storage slot for a large contact CSV
/// (POST /audiences/:id/contacts/import/upload). PUT the file to
/// <see cref="UploadUrl"/> with <c>Content-Type: </c><see cref="ContentType"/>,
/// then pass <see cref="StorageKey"/> to ContactImportStorageKeyAsync.
/// </summary>
public class ContactImportUpload
{
    [JsonPropertyName("object")]
    public string Object { get; set; } = "contact_import_upload";

    /// <summary>Hand this back to the import endpoint once the upload finished.</summary>
    [JsonPropertyName("storage_key")]
    public string StorageKey { get; set; } = null!;

    /// <summary>A bearer credential in URL form — never log it.</summary>
    [JsonPropertyName("upload_url")]
    public string UploadUrl { get; set; } = null!;

    /// <summary>The Content-Type the presigned URL was signed for.</summary>
    [JsonPropertyName("content_type")]
    public string? ContentType { get; set; }

    /// <summary>Seconds until <see cref="UploadUrl"/> expires.</summary>
    [JsonPropertyName("expires_in")]
    public int ExpiresIn { get; set; }

    /// <summary>Upload size ceiling in bytes (256 MB).</summary>
    [JsonPropertyName("max_bytes")]
    public long MaxBytes { get; set; }
}

/// <summary>Options for listing contacts (limit ≤ 100).</summary>
public class ContactListOptions : PaginationOptions
{
    /// <summary>
    /// The sending domain whose contact pool to list (domain-first model).
    /// REQUIRED on the flat /contacts API (whenever <see cref="AudienceId"/> is omitted).
    /// </summary>
    public string? Domain { get; set; }

    /// <summary>List a specific audience via the nested /audiences/:id/contacts API instead of <see cref="Domain"/>.</summary>
    public string? AudienceId { get; set; }

    /// <summary>Restrict to members of this segment (works with or without AudienceId).</summary>
    public string? SegmentId { get; set; }
}

/// <summary>Delete acknowledgement for a contact: <c>{ object: 'contact', id, deleted }</c>.</summary>
public class ContactDeleted
{
    [JsonPropertyName("object")]
    public string Object { get; set; } = "contact";

    /// <summary>The deleted contact's id.</summary>
    [JsonPropertyName("id")]
    public string Id { get; set; } = null!;

    /// <summary>The deleted contact's id.</summary>
    [Obsolete("The API returns the id under `id`, never `contact`. Use Id instead; this alias will be removed in 2.0.")]
    [JsonIgnore]
    public string Contact => Id;

    [JsonPropertyName("deleted")]
    public bool Deleted { get; set; }
}

/// <summary>Acknowledgement of removing a contact from a segment.</summary>
public class SegmentMembershipRemoved
{
    [JsonPropertyName("id")]
    public string Id { get; set; } = null!;

    [JsonPropertyName("audienceId")]
    public string? AudienceId { get; set; }

    [JsonPropertyName("deleted")]
    public bool Deleted { get; set; }
}

/// <summary>One topic subscription state of a contact.</summary>
public class ContactTopicSubscription
{
    [JsonPropertyName("id")]
    public string Id { get; set; } = null!;

    [JsonPropertyName("name")]
    public string Name { get; set; } = null!;

    [JsonPropertyName("description")]
    public string? Description { get; set; }

    /// <summary><c>opt_in</c> | <c>opt_out</c>.</summary>
    [JsonPropertyName("subscription")]
    public string Subscription { get; set; } = null!;
}

/// <summary>A contact's topic subscriptions (GET /contacts/:id/topics).</summary>
public class ContactTopics
{
    [JsonPropertyName("object")]
    public string Object { get; set; } = "list";

    /// <summary>
    /// Whether another page exists. The endpoint only pages when
    /// ContactRetrieveTopicsAsync is called with a <see cref="PaginationOptions"/>.
    /// </summary>
    [JsonPropertyName("has_more")]
    public bool HasMore { get; set; }

    [JsonPropertyName("data")]
    public List<ContactTopicSubscription> Data { get; set; } = new();
}

/// <summary>One topic setting in a topics update.</summary>
public class ContactTopicSetting
{
    [JsonPropertyName("id")]
    public string Id { get; set; } = null!;

    /// <summary><c>opt_in</c> | <c>opt_out</c>.</summary>
    [JsonPropertyName("subscription")]
    public string Subscription { get; set; } = null!;
}

/// <summary>Payload for PATCH /contacts/:id/topics.</summary>
public class ContactTopicsUpdateOptions
{
    [JsonPropertyName("topics")]
    public List<ContactTopicSetting> Topics { get; set; } = new();
}

/// <summary>A registered custom contact property (merge tag).</summary>
public class ContactProperty
{
    [JsonPropertyName("object")]
    public string Object { get; set; } = "contact_property";

    [JsonPropertyName("id")]
    public string Id { get; set; } = null!;

    [JsonPropertyName("key")]
    public string Key { get; set; } = null!;

    /// <summary><c>string</c> | <c>number</c>.</summary>
    [JsonPropertyName("type")]
    public string Type { get; set; } = null!;

    /// <summary>String, number or null.</summary>
    [JsonPropertyName("fallback_value")]
    public JsonElement? FallbackValue { get; set; }

    [JsonPropertyName("created_at")]
    public string CreatedAt { get; set; } = null!;
}

/// <summary>Payload for registering a custom contact property (POST /contact-properties).</summary>
public class ContactPropertyCreateOptions
{
    /// <summary>Canonical merge-tag key. <see cref="Name"/> is accepted as an alias.</summary>
    [JsonPropertyName("key")]
    public string? Key { get; set; }

    /// <summary>Alias for <see cref="Key"/>.</summary>
    [JsonPropertyName("name")]
    public string? Name { get; set; }

    /// <summary><c>string</c> | <c>number</c>. Required.</summary>
    [JsonPropertyName("type")]
    public string Type { get; set; } = null!;

    /// <summary>Default merge value when a contact has none (string or number).</summary>
    [JsonPropertyName("fallback_value")]
    public object? FallbackValue { get; set; }
}
